content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-3.md

```diff
@@ -0,0 +1,295 @@
+
+![MapR Platform](/img/mapr-platform.png)
+
+This is the third in a series of blogs, which discusses the architecture of a data pipeline that combines streaming data with machine learning and fast storage. [The first post](https://developer.hpe.com/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-/) discussed creating a machine learning model to predict flight delays. The [second post](https://developer.hpe.com/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-2/) discussed using the saved model with streaming data to do real-time analysis of flight delays. This third post will discuss fast storage and analysis with MapR Database, Apache Spark, Apache Drill and OJAI.
+
+**Machine Learning Logistics and Data Pipelines**
+
+Machine Learning usually refers to the model training piece of an ML workflow. But, as Ted Dunning says, 90% of the effort around Machine Learning is data logistics which includes all of the aspects that occur before and after this training.  When you combine event streams with microservices, you can greatly enhance the agility with which you build, deploy, and maintain complex data pipelines. Pipelines are constructed by chaining together microservices, each of which listens for the arrival of some data, performs its designated task, and optionally publishes its own messages to another topic.  Combining event-driven data pipelines with machine learning can handle the logistics of machine learning in a flexible way by:
+
+* Making input and output data available to independent consumers
+* Managing and evaluating multiple models and easily deploying new models
+* Monitoring and analyzing models, with historical and real-time data.
+
+![MapR Architectures](/img/mapr-architectures.png)
+
+Architectures for these types of applications are discussed in more detail in the eBooks: Machine Learning Logistics, Streaming Architecture, and Microservices and Containers.
+
+The following figure depicts the (simplified) data pipeline for this tutorial:
+
+* Flight trip data is published to a MapR Event Streams (ES) topic using the Kafka API. (Note this data contains the actual delay label, in the real world architecture the actual delay label would come later in a different topic, but to keep the tutorial code simple it is combined with the input data).
+* A Spark Streaming application subscribed to the first topic enriches the event with the flight predictions and publishes the results in JSON format to another topic. ( In the real world architecture there would be multiple consumers publishing model predictions, but to keep the tutorial code simple there is only one here).
```

Comment:
Add hyphen between “real” and “world” 
Add a comma after “architecture”

![image](https://user-images.githubusercontent.com/77496261/137183169-cf4d7784-3e56-42aa-af56-ef634b6221f7.png)

Reply:
done